### ISSUE TITLE
Implement Dual Quaternion Skinning AKA "Preserve Volume" option compatibility with Blender

### ISSUE THREAD

**Issue by beiller** [godotengine/godot-proposals#9235]
### Describe the project you are working on

Working on a character importing and posing application where the user can use VR to import / export poses and animations.

### Describe the problem or limitation you are having in your project

There is currently no compatibility in Godot with blenders "preserve volume" option on Skeletons

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

With some small changes to the Skeleton3D pipeline Dual Quaternion can be implemented and offer better compatibility with Blender.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Initial implementation PR is here:
https://github.com/godotengine/godot/pull/89131

- Dual Quaternion transform needs to be generated from the matrix transform per-bone.
- 2 Quaternions need to be passed into the vertex() program, and blended together (x4 or x8)
- The final transform is computed in vertex() program in mat4 format and applied to the vertex

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can be implemented in godot via gdscript / shader. That would require the complete implementation of BasicMaterial in glsl and is not easy to carry on through version changes etc.

### Is there a reason why this should be core and not an add-on in the asset library?

As above, it requires very significant effort to write a ShaderMaterial and to keep that material in line with existing BasicMaterial seems to be very difficult / impossible, and the hypothetical add-on written would not be compatible with any minor changes in the engine going forward. The change as presented (in-core) is much more portable and should survive major refactors.

**Comment by beiller** [comment]
Hey its not correct see the linked PR for a full working implementation. It just involves converting the bone transform mat4 to 2 quaternions, passing those quaternions to the existing shader, and in the shader blending the quaternions and converting back to matrix 4. No data changes at all required.

**Comment by clayjohn** [comment]
> Hey its not correct see the linked PR for a full working implementation. It just involves converting the bone transform mat4 to 2 quaternions, passing those quaternions to the existing shader, and in the shader blending the quaternions and converting back to matrix 4. No data changes at all required.

Is it correct to say that the benefit of this approach is just the lowered bandwidth requirement? And the tradeoff is increased instruction count both inside the shader and on CPU?

**Comment by beiller** [comment]
Hey @clayjohn  the benefit of dual quaternion transforms for Skeletons is that twisting a skeletal joint causes the mesh to collapse under linear transformation. Using the quaternion blending "preserves the volume" during rotation. You can see the difference in blender using an armature and bending it like a knee joint and toggle on/off "preserve volume". In linear space the faces will collapse into very thin areas and the DQ transform will preserve the volume of the joint. It's used in a lot of character rigs for joints for that reason.

**Comment by TokageItLab** [comment]
When a skin is linearly deformed, the large curvatures cause the skin to lose its shape. This is usually solved with [BoneConstraint and subsidiary bones](https://iloverisusan.tumblr.com/post/667182220317999104/%EF%BC%92%E3%81%A4%E3%81%AE%E9%AA%A8%E3%81%AE%E5%BD%B1%E9%9F%BF%E3%82%92%E5%8F%97%E3%81%91%E3%82%8B%E8%A3%9C%E5%8A%A9%E9%AA%A8%E3%82%92%E4%BD%9C%E3%82%8B), but DualQuaternion serves as [ another solution](https://users.cs.utah.edu/~ladislav/dq/index.html). BoneConstraint is still a work in progress and will be implemented as the next step after BoneModifier.

Dual Quaternion will improve the appearance of skinning in some cases, although it will be twice as computationally intensive since two Quaternions will be needed for each calculation.

The problem is that the skin will not lose weight, but on the other hand, it may expand unnaturally, making it unsuitable for low-polygon models. Also, as mentioned above, many animations are resolved with auxiliary bones, so this feature is not required.

I am not opposed to including this feature in core because it would be difficult to make it an extension, but if it is implemented, it needs to be optional. Also, a warning should be given that this skinning setting is generally ignored in glTF Export, as it is platform-dependent whether this calculation is done for skinning by similar reason for NLA animation.

**Comment by beiller** [comment]
@TokageItLab nice thanks for the writeup. Low poly definitely sounds like it could cause some problems. Not sure how bad they would be, but on my TODO in the PR is a toggle it should not be difficult to implement on Skeleton3D object.

For the computation modern graphics card has no issue. The technique seems to come from around 2006. I think theoretically the bandwidth savings is worth more than computation cost. But we currently do not even take advantage of lower bandwidth since in my current PR the quaternions are packed into the matrix.

**Comment by TokageItLab** [comment]
> Not sure how bad they would be, but on my TODO in the PR is a toggle it should not be difficult to implement on Skeleton3D object.

For example, if you want skinning methods to coexist on the skin and armor parts of a hard-surface robot, I think it would be more preferable to have a toggle on the MeshInstance side.

**Comment by fire** [comment]
For example, a newer technique is called delta mush https://github.com/V-Sekai/unity-mesh-deform-direct-delta-mush and I agree with @TokageItLab that the code can be worth abstracting between linear blending and dual quaternion blending.

**Comment by beiller** [comment]
I think that to accomplish these other things, the vertex pipeline just needs to be programmable in Godot similar to shaders. Currently the vertex pipeline is a compute shader but it can't be swapped out in the editor. That should be the future direction imo.

Setting dual quaternion use per mesh instance is tricky as well it means we have to be less efficient and compute both dual quaternion and linear blend matrices always or said another way, all skeleton has to compute dual quaternion. Which maybe isn't a big deal.

**Comment by TokageItLab** [comment]
I think the performance loss is simply a matter of architectural design and can be avoided by using a cache. If there is at least one DQS, the calculation for the DQS can be cached and used for multiple MeshInstances, the same goes for LinearSkinning. In other words, if both Skinning coexist.be used by MeshInstances, there will be two caches, and if only one of them is used, there will be one cache.

**Comment by octanejohn** [comment]
related i think

https://github.com/enkimute/LookMaNoMatrices

https://enkimute.github.io/LookMaNoMatrices/

**Comment by firebeetleLobster** [comment]
Try this: https://godotengine.org/asset-library/asset/2062

**Comment by computerologist** [comment]
> BTW, since there are several other possible methods of skinning besides DualQuaternion, it would be desirable if the implementation way could be extended to them in the future. Well, boolean is fine for now even if we make it an enum in the future.

Superior to this would be a weighted blend of techniques, per vertex. It is inherently inefficient, but allows artists the option to trade efficiency for maximally localised quality across a mesh--the typical example of this is in the DQS vs LBS optimizations of candywrappering of forearms vs inflation of elbows.

**Comment by beiller** [comment]
Yes I tried this approach. When you write a custom shader now all the spatial shader techniques are gone unless you rewrite them yourself. It's very difficult to use in practice I found.

**Comment by Glitshy** [comment]
> > BTW, since there are several other possible methods of skinning besides DualQuaternion, it would be desirable if the implementation way could be extended to them in the future. Well, boolean is fine for now even if we make it an enum in the future.
> 
> Superior to this would be a weighted blend of techniques, per vertex. It is inherently inefficient, but allows artists the option to trade efficiency for maximally localised quality across a mesh--the typical example of this is in the DQS vs LBS optimizations of candywrappering of forearms vs inflation of elbows.

I wonder how this information (i.e. which skinning method to use) would be stored in the mesh. Is there already any other engine or file type (fbx, glb) using an approach like this, or would this have to be stored as custom vertex data?

Also, if we assume that only DQS and LBS will be implemented, then a single float per vertice is sufficient to interpolate between the two. But if there are multiple available skinning methods (as suggested by others in this issue), should there then be one float per vertice and per available method (similar to weight painting I guess with one weight per bone)?

Lastly, does anyone know if implementing DQS is actively being worked on? I have just imported my character from blender and the forearm is candywrappering, so now I have to look into work arounds :/

**Comment by fire** [comment]
I am somewhat interested in delta mush https://github.com/V-Sekai/unity-mesh-deform-direct-delta-mush
